### PR TITLE
Add option to use GPU

### PR DESCRIPTION
**Pull request by ayush-1506**
Fixes: #57

### CODE REVIEW

**Comment by ablaom** [comment]
Perhaps for consistency with MLJ we should use https://github.com/timholy/ComputationalResources.jl and the hyper-parameter should be called `acceleration`? See https://alan-turing-institute.github.io/MLJ.jl/dev/acceleration_and_parallelism/

Would MLJFlux training with `acceleration=CUDALibs()` play well with MLJ `acceleration` options for `evaluate!` and tuning?

@OkonSamuel Be good to have your input,  maybe even a code review, if you have time.

File: src/classifier.jl [hunk 1]
```diff
@@ -8,6 +8,7 @@ mutable struct NeuralNetworkClassifier{B,F,O,L} <: MLJModelInterface.Probabilist
     lambda::Float64 # regularization strength
     alpha::Float64  # regularizaton mix (0 for all l2, 1 for all l1)
     optimiser_changes_trigger_retraining::Bool
+    gpu::Bool       # Train on gpu
```
Comment (OkonSamuel) [inline src/classifier.jl:11]: Let make things consistent as @ablaom as suggested by making use of `ComputationalResources.jl`. Maybe something in the lines of.
```julia
function use_gpu(resource)
 if resource isa Cudalibs
     return true
else if resource isa CPU1
      return false
else
      throw(ArgumentError("Resource not supported"))
end
end#you might want to create a `clean!` method

mutable struct NeuralNetworkClassifier{B,F,O,L} <: MLJModelInterface.Probabilistic
    builder::B
    finaliser::F
    optimiser::O    # mutable struct from Flux/src/optimise/optimisers.jl
    loss::L         # can be called as in `loss(yhat, y)`
    epochs::Int          # number of epochs
    batch_size::Int # size of a batch
    lambda::Float64 # regularization strength
    alpha::Float64  # regularizaton mix (0 for all l2, 1 for all l1)
    optimiser_changes_trigger_retraining::Bool
    gpu::AbstractResource       # Train on gpu
end

NeuralNetworkClassifier(; builder::B   = Short()
              , finaliser::F = Flux.softmax
              , optimiser::O = Flux.Optimise.ADAM()
              , loss::L      = Flux.crossentropy
              , epochs       = 10
              , batch_size   = 1
              , lambda       = 0
              , alpha        = 0
              , optimiser_changes_trigger_retraining = false
              , gpu  = CPU1()
              ) where {B,F,O,L} =
                  NeuralNetworkClassifier{B,F,O,L}(builder
                                       , finaliser
                                       , optimiser
                                       , loss
                                       , epochs
                                       , batch_size
                                       , lambda
                                       , alpha
                                       , optimiser_changes_trigger_retraining
                                       , gpu
                                       )

```
then 
```julia
chain, history = fit!(chain, optimiser, model.loss,
                          model.epochs, model.lambda,
                          model.alpha, verbosity, data, model.gpu)
```
becomes
```julia
chain, history = fit!(chain, optimiser, model.loss,
                          model.epochs, model.lambda,
                          model.alpha, verbosity, data, use_gpu(model.gpu))
```

**Review by OkonSamuel — CHANGES_REQUESTED** [review]

**Comment by OkonSamuel** [comment]
@ayush-1506 Thanks. This is awesome.!!!
@ablaom i don't see why this won't with `evaluate` or tuning (We need to test this to make sure there are no glitches).
The only issue i can foresee is when using `Cudalibs()` with `CPUThreads()` in `evaluate` and we should be able to modify code at MLJBase to catch that case.

**Comment by ablaom** [comment]
@OkonSamuel Thanks. This looks good. @ayush-1506 What do you think?

> The only issue i can foresee is when using Cudalibs() with CPUThreads() in evaluate and we should be able to modify code at MLJBase to catch that case.

👍 

> (We need to test this to make sure there are no glitches).

 I'm going to rebase this PR onto the dev branch for testing purposes.

**Comment by ayush-1506** [comment]
Thanks for the review @OkonSamuel ! The changes should solve this issue. If someone had GPU access, would love your thoughts.

cc @ablaom

File: src/core.jl [hunk 2]
```diff
@@ -86,6 +96,12 @@ function  fit!(chain, optimiser, loss, epochs,
     loss_func(x, y) = loss(chain(x), y)
     history = []
     prev_loss = Inf
+    if gpu
+        @info "Using GPU for training"
+        data = Flux.gpu.(data)
```
Comment (ablaom) [inline src/core.jl:101]: Suggest suppressing the `@info` for `verbosity < 1`.

**Review by ablaom — COMMENTED** [review]
Just that one small detail. I'll merge this onto `dev` for now. 

Am looking into getting a `GPU` for one-time testing. 

@OkonSamuel If you have time, would be good to throw together a test script. It could live in this repo even if we don't add it to CI.

File: src/core.jl (hunk 2 above)
Comment (ablaom) [reply src/core.jl:101]: Shouldn't this be `verbosity <  1 || @info "Using GPU for training` or `verbosity > 0 && @info "Using GPU for training"`.?

Comment (ayush-1506) [reply src/core.jl:101]: fixed, my bad.

**Review by ablaom — APPROVED** [review]
👍

**Pull request merged**